cpp/src/toppra/algorithm.cpp

```diff
@@ -67,6 +86,100 @@ ReturnCode PathParametrizationAlgorithm::computeControllableSets(
   return ret;
 }
 
+ReturnCode PathParametrizationAlgorithm::computePathParametrizationParallel(value_type vel_start,
+                                                                            value_type vel_end) {
+
+  size_t be = GetCurrentTimeUsec();
+  initialize();
+  size_t en = GetCurrentTimeUsec();
+  printf("time of init:                             %fms\n", (en - be)/1000.0);
+  be = en;
+
+  m_solver->setupSolver();
+  Bound vel_ends;
+  vel_ends.setConstant(vel_end);
+  m_data.ret_code = computeControllableSetsParallel(vel_ends);
+  en = GetCurrentTimeUsec();
+  printf("time of computeControllableSetsParallel:  %fms\n", (en - be)/1000.0);
+  be = en;
+  if ((int)m_data.ret_code > 0) {
+    return m_data.ret_code;
+  }
+
+  m_data.ret_code = computeForwardPass(vel_start);
+  en = GetCurrentTimeUsec();
+  printf("time of computeForwardPass:               %fms\n", (en - be)/1000.0);
+  return m_data.ret_code;
+};
+
+ReturnCode PathParametrizationAlgorithm::computeControllableSetsParallel(
+    const Bound &vel_ends) {
+  ReturnCode ret = ReturnCode::OK;
+  bool solver_ret;
+  Vector g_upper{2}, g_lower{2}, solution;
+  // Vectors  solution_upper, solution_lower;
+  // solution_upper.assign(m_N, Vector::Zero(2, 1));
+  // solution_lower.assign(m_N, Vector::Zero(2, 1));
+
+  g_upper << 1e-9, -1;
+  g_lower << -1e-9, 1;
+  m_data.controllable_sets.row(m_N) = vel_ends.cwiseAbs2();
+
+  size_t be = GetCurrentTimeUsec();
+
+  bool batch_ok = true;
+  printf("omp omp_get_max_threads %d\n", omp_get_max_threads());
+  #pragma omp parallel for schedule(dynamic)
+  for (int i = 0; i < m_N; i++) {
+    // solver_ret = m_solver->solveStagewiseBatch(i, g_upper, solution_upper[i]);
+    solver_ret = m_solver->solveStagewiseBatch(i, g_upper);
```

Comment:
This doesn't look right to me. Are you sure that this routine returns the correct controllable sets?

Reply:
No, it just store the half-finished solution in solver.

Reply:
And here is actually the implementation of parallelism.